src/jit/lowerarmarch.cpp

```diff
@@ -704,7 +704,35 @@ void Lowering::ContainCheckCast(GenTreeCast* node)
 //
 void Lowering::ContainCheckCompare(GenTreeOp* cmp)
 {
-    ContainCheckBinary(cmp);
+    if (CheckImmedAndMakeContained(cmp, cmp->gtOp2))
+    {
+#ifdef _TARGET_ARM64_
+        GenTreePtr op1 = cmp->gtOp.gtOp1;
+        GenTreePtr op2 = cmp->gtOp.gtOp2;
+
+        // If op1 codegen can set flags op2 is an immediate 0
+        // we don't need to generate cmp instruction,
+        // provided we don't have another GenTree node between op1
+        // and cmp that could potentially modify flags.
+        //
+        // TODO-CQ: right now the below peep is inexpensive and
+        // gets the benefit in most of cases because in majority
+        // of cases op1, op2 and cmp would be in that order in
+        // execution.  In general we should be able to check that all
+        // the nodes that come after op1 in execution order do not
+        // modify the flags so that it is safe to avoid generating a
+        // test instruction.  Such a check requires that on each
+        // GenTree node we need to set the info whether its codegen
+        // will modify flags.
+        if (op2->IsIntegralConst(0) && (op1->gtNext == op2) && (op2->gtNext == cmp) &&
+            cmp->OperIs(GT_EQ, GT_NE, GT_GT, GT_GE, GT_LT, GT_LE) && op1->OperIs(GT_ADD, GT_AND, GT_SUB))
```

Comment:
ARM64 can only set flags on `adc`, `add`, `and`, `bics`, `sub`, `sbc` (and their aliases `cmp`, `cmn`, `tst` ...).

Reply:
So this is the only difference between the ARM64 version and the XARCH version. It really should be in `LowerCompare`, with or without my refactoring.

Reply:
This PR?  Or a subsequent PR?

Reply:
Given the very restrictive set of ops which can set flags, it may be easier to relax the 
`&& (op1->gtNext == op2) && (op2->gtNext == cmp) &&` restriction.

Also `cbz`, `cbnz`, `tbz`, `tbnz` will also change things.

Reply:
I guess my preference would be to refactor a bit later.  (Next week?)

Reply:
> Given the very restrictive set of ops which can set flags, it may be easier to relax the
&& (op1->gtNext == op2) && (op2->gtNext == cmp) && restriction.

Right, unlike on xarch few instruction set flags and only when the S version is used. On xarch we need to be more careful because most typical instructions do set the flags.

Reply:
> This PR? Or a subsequent PR?

IMO subsequent. Let's see what happens with my proposed refactoring